File: backend/utils/sessionStore.js
const logger = require("./logger");

class SessionStore {
  constructor() {
    this.sessions = new Map();
    this.cleanupInterval = 30 * 60 * 1000; // 30分钟清理一次过期会话
    this.sessionTimeout = 2 * 60 * 60 * 1000; // 2小时会话超时
    this.confirmationTimeout = 5 * 60 * 1000; // 5分钟确认超时

    this.startCleanupTimer();
  }

  // 创建或获取会话
  getOrCreateSession(socketId) {
    if (!this.sessions.has(socketId)) {
      this.sessions.set(socketId, {
        sessionId: socketId,
        history: [],
        pendingToolCall: null,
        pendingConfirmation: null,
        canceled: false,
        createdAt: new Date(),
        lastActivity: new Date(),
        context: {},
      });

      logger.info(`创建新会话: ${socketId}`);
    }

    const session = this.sessions.get(socketId);
    session.lastActivity = new Date();

    return session;
  }

  // 添加消息到历史记录
  addMessage(socketId, message) {
    const session = this.getOrCreateSession(socketId);

    const messageRecord = {
      id: `msg_${Date.now()}_${Math.random().toString(36).substr(2, 9)}`,
      timestamp: new Date(),
      type: message.type, // 'user', 'assistant', 'system', 'tool_call', 'tool_result'
      content: message.content,
      metadata: message.metadata || {},
    };

    session.history.push(messageRecord);

    // 限制历史记录长度，保留最近100条
    if (session.history.length > 100) {
      session.history = session.history.slice(-100);
    }

    logger.debug(`添加消息到会话 ${socketId}`, {
      messageId: messageRecord.id,
      type: messageRecord.type,
    });

    return messageRecord;
  }

  // 设置待处理的工具调用
  setPendingToolCall(socketId, toolCall, riskAssessment = null) {
    const session = this.getOrCreateSession(socketId);

    session.pendingToolCall = {
      ...toolCall,
      riskAssessment: riskAssessment,
      createdAt: new Date(),
      expiresAt: new Date(Date.now() + this.confirmationTimeout),
    };

    session.canceled = false;

    logger.info(`设置待处理工具调用: ${socketId}`, {
      toolName: toolCall.name,
      requiresConfirmation: riskAssessment?.requiresConfirmation || false,
    });

    return session.pendingToolCall;
  }

  // 清除待处理的工具调用
  clearPendingToolCall(socketId) {
    const session = this.sessions.get(socketId);
    if (session) {
      const pendingCall = session.pendingToolCall;
      session.pendingToolCall = null;
      session.canceled = false;

      if (pendingCall) {
        logger.info(`清除待处理工具调用: ${socketId}`, {
          toolName: pendingCall.name,
        });
      }

      return pendingCall;
    }

    return null;
  }

  // 标记工具调用为已取消
  markCanceled(socketId, reason = "user_canceled") {
    const session = this.sessions.get(socketId);
    if (session) {
      session.canceled = true;
      session.cancelReason = reason;
      session.canceledAt = new Date();

      logger.info(`标记工具调用已取消: ${socketId}`, { reason });

      return true;
    }

    return false;
  }

  // 设置待确认的请求
  setPendingConfirmation(socketId, confirmationRequest) {
    const session = this.getOrCreateSession(socketId);

    session.pendingConfirmation = {
      ...confirmationRequest,
      createdAt: new Date(),
      expiresAt: new Date(Date.now() + this.confirmationTimeout),
    };

    logger.info(`设置待确认请求: ${socketId}`, {
      confirmationId: confirmationRequest.id,
      toolName: confirmationRequest.toolCall.name,
    });

    return session.pendingConfirmation;
  }

  // 获取待确认的请求
  getPendingConfirmation(socketId) {
    const session = this.sessions.get(socketId);
    if (session?.pendingConfirmation) {
      // 检查是否过期
      if (new Date() > session.pendingConfirmation.expiresAt) {
        this.clearPendingConfirmation(socketId);
        return null;
      }

      return session.pendingConfirmation;
    }

    return null;
  }

  // 清除待确认的请求
  clearPendingConfirmation(socketId) {
    const session = this.sessions.get(socketId);
    if (session) {
      const confirmation = session.pendingConfirmation;
      session.pendingConfirmation = null;

      if (confirmation) {
        logger.info(`清除待确认请求: ${socketId}`, {
          confirmationId: confirmation.id,
        });
      }

      return confirmation;
    }

    return null;
  }

  // 获取会话历史记录
  getHistory(socketId, limit = 50) {
    const session = this.sessions.get(socketId);
    if (session) {
      return session.history.slice(-limit);
    }

    return [];
  }

  // 获取会话上下文
  getContext(socketId) {
    const session = this.sessions.get(socketId);
    return session ? session.context : {};
  }

  // 更新会话上下文
  updateContext(socketId, contextUpdate) {
    const session = this.getOrCreateSession(socketId);
    session.context = { ...session.context, ...contextUpdate };

    logger.debug(`更新会话上下文: ${socketId}`, contextUpdate);

    return session.context;
  }

  // 获取会话状态
  getSessionStatus(socketId) {
    const session = this.sessions.get(socketId);
    if (!session) {
      return {
        exists: false,
        active: false,
      };
    }

    const now = new Date();
    const isExpired = now - session.lastActivity > this.sessionTimeout;

    return {
      exists: true,
      active: !isExpired,
      sessionId: session.sessionId,
      hasPendingToolCall: !!session.pendingToolCall,
      hasPendingConfirmation: !!session.pendingConfirmation,
      isCanceled: session.canceled,
      messageCount: session.history.length,
      lastActivity: session.lastActivity,
      createdAt: session.createdAt,
    };
  }

  // 清理过期会话
  cleanupExpiredSessions() {
    const now = new Date();
    const expiredSessions = [];

    for (const [socketId, session] of this.sessions.entries()) {
      const isExpired = now - session.lastActivity > this.sessionTimeout;
      const hasExpiredConfirmation =
        session.pendingConfirmation && now > session.pendingConfirmation.expiresAt;
      const hasExpiredToolCall = session.pendingToolCall && now > session.pendingToolCall.expiresAt;

      if (isExpired || hasExpiredConfirmation || hasExpiredToolCall) {
        expiredSessions.push(socketId);
      }
    }

    for (const socketId of expiredSessions) {
      const session = this.sessions.get(socketId);
      logger.info(`清理过期会话: ${socketId}`, {
        expiredReason: session ? "session_timeout" : "not_found",
        lastActivity: session?.lastActivity,
      });

      this.sessions.delete(socketId);
    }

    if (expiredSessions.length > 0) {
      logger.info(`清理完成，删除 ${expiredSessions.length} 个过期会话`);
    }

    return expiredSessions.length;
  }

  // 启动清理定时器
  startCleanupTimer() {
    setInterval(() => {
      this.cleanupExpiredSessions();
    }, this.cleanupInterval);

    logger.info("会话清理定时器已启动", {
      interval: this.cleanupInterval,
      timeout: this.sessionTimeout,
    });
  }

  // 获取所有活跃会话统计
  getStats() {
    const now = new Date();
    let activeCount = 0;
    let pendingToolCalls = 0;
    let pendingConfirmations = 0;
    let totalMessages = 0;

    for (const session of this.sessions.values()) {
      const isActive = now - session.lastActivity < this.sessionTimeout;
      if (isActive) activeCount++;

      if (session.pendingToolCall) pendingToolCalls++;
      if (session.pendingConfirmation) pendingConfirmations++;
      totalMessages += session.history.length;
    }

    return {
      totalSessions: this.sessions.size,
      activeSessions: activeCount,
      pendingToolCalls: pendingToolCalls,
      pendingConfirmations: pendingConfirmations,
      totalMessages: totalMessages,
      averageMessagesPerSession:
        this.sessions.size > 0 ? Math.round(totalMessages / this.sessions.size) : 0,
    };
  }

  // 手动清理指定会话
  clearSession(socketId) {
    const session = this.sessions.get(socketId);
    if (session) {
      logger.info(`手动清理会话: ${socketId}`, {
        messageCount: session.history.length,
        hasPendingToolCall: !!session.pendingToolCall,
        hasPendingConfirmation: !!session.pendingConfirmation,
      });

      this.sessions.delete(socketId);
      return true;
    }

    return false;
  }

  // 导出会话数据（用于调试）
  exportSession(socketId) {
    const session = this.sessions.get(socketId);
    if (session) {
      return {
        sessionId: session.sessionId,
        history: session.history,
        context: session.context,
        pendingToolCall: session.pendingToolCall,
        pendingConfirmation: session.pendingConfirmation,
        status: this.getSessionStatus(socketId),
      };
    }

    return null;
  }

  // 获取所有会话ID（用于管理）
  getAllSessionIds() {
    return Array.from(this.sessions.keys());
  }
}

module.exports = SessionStore;

File: backend_py/session_store.py
from __future__ import annotations

import time
from dataclasses import dataclass, field
from datetime import datetime, timedelta
from typing import Any, Dict, List, Optional


@dataclass
class Session:
    session_id: str
    history: List[Dict[str, Any]] = field(default_factory=list)
    pending_tool_call: Optional[Dict[str, Any]] = None
    pending_confirmation: Optional[Dict[str, Any]] = None
    canceled: bool = False
    created_at: datetime = field(default_factory=datetime.utcnow)
    last_activity: datetime = field(default_factory=datetime.utcnow)
    context: Dict[str, Any] = field(default_factory=dict)

    # 运行时标志位
    tts_settings: Dict[str, Any] = field(default_factory=dict)
    allow_local_control: Optional[bool] = None
    network_access_enabled: bool = False
    # 最近一次“显式设置联网开关”的时间戳（ms）。用于 sid->clientId 迁移合并时判定新旧优先级。
    network_access_set_at_ms: int = 0
    device_location_enabled: bool = False
    # 最近一次“显式设置设备定位开关”的时间戳（ms）。用于 sid->clientId 迁移合并时判定新旧优先级。
    device_location_set_at_ms: int = 0
    device_location: Optional[Dict[str, Any]] = None
    tts_stopped: bool = False
    current_request_id: Optional[str] = None


class SessionStore:
    """内存会话存储（与 backend/utils/sessionStore.js 行为对齐）。"""

    def __init__(self) -> None:
        self._sessions: Dict[str, Session] = {}
        self.session_timeout = timedelta(hours=2)
        self.confirmation_timeout = timedelta(minutes=5)

    @staticmethod
    def _merge_session_state(*, dst: Session, src: Session) -> None:
        """将 src 的“会话态”合并进 dst（最小化覆盖）。

        设计原则（用于修复 sid->clientId 迁移的时序竞争）：
        - 不再在 dst 已存在时直接丢弃 src（会导致开关/音色等丢失）
        - 只合并“稳定配置/权限/历史”等字段，避免覆盖高风险的 pending 状态
        - 对布尔开关采用“显式设置时间戳优先”（既能保留开启，也能正确传播关闭）
        - 对 Optional 字段采用“dst 缺省时用 src”（防止误覆盖已有配置）
        """

        # 1) 权限/开关：按“最近一次显式设置”的时间戳合并（支持开启与关闭都能正确传播）。
        dst_net_ts = int(getattr(dst, "network_access_set_at_ms", 0) or 0)
        src_net_ts = int(getattr(src, "network_access_set_at_ms", 0) or 0)
        if src_net_ts > dst_net_ts:
            dst.network_access_enabled = bool(src.network_access_enabled)
            dst.network_access_set_at_ms = src_net_ts

        dst_loc_ts = int(getattr(dst, "device_location_set_at_ms", 0) or 0)
        src_loc_ts = int(getattr(src, "device_location_set_at_ms", 0) or 0)
        if src_loc_ts > dst_loc_ts:
            dst.device_location_enabled = bool(src.device_location_enabled)
            dst.device_location_set_at_ms = src_loc_ts

        # 2) 设备定位缓存（仅当 dst 缺失时使用 src）
        if dst.device_location is None and isinstance(src.device_location, dict):
            dst.device_location = src.device_location
        # 若 src 的设备定位“显式设置”更新更晚，则允许覆盖（例如用户关闭定位时需要清空缓存）
        if src_loc_ts > dst_loc_ts:
            dst.device_location = src.device_location if isinstance(src.device_location, dict) else None

        # 3) allow_local_control（仅当 dst 未显式设置时使用 src）
        if dst.allow_local_control is None and isinstance(src.allow_local_control, bool):
            dst.allow_local_control = src.allow_local_control

        # 4) tts_settings（对缺失键做补齐，不覆盖已有键）
        if isinstance(src.tts_settings, dict) and src.tts_settings:
            if not isinstance(dst.tts_settings, dict):
                dst.tts_settings = {}
            for k, v in src.tts_settings.items():
                if k not in dst.tts_settings:
                    dst.tts_settings[k] = v
                    continue

                existing = dst.tts_settings.get(k)
                is_empty_dict = isinstance(existing, dict) and (not existing)
                if existing is None or existing == "" or is_empty_dict:
                    dst.tts_settings[k] = v

        # 5) history（合并并截断到最后 100 条）
        if isinstance(src.history, list) and src.history:
            if not isinstance(dst.history, list):
                dst.history = []
            dst.history.extend([x for x in src.history if isinstance(x, dict)])
            if len(dst.history) > 100:
                dst.history = dst.history[-100:]

        # 6) requestId（仅当 dst 缺失时使用 src）
        if (dst.current_request_id is None) and isinstance(src.current_request_id, str) and src.current_request_id.strip():
            dst.current_request_id = src.current_request_id

        # 7) 生命周期字段：created_at 取更早的，last_activity 取更晚的
        try:
            dst.created_at = min(dst.created_at, src.created_at)
        except Exception:
            pass
        try:
            dst.last_activity = max(dst.last_activity, src.last_activity)
        except Exception:
            pass

    def migrate(self, from_id: str, to_id: str) -> None:
        """将会话从一个 id 迁移到另一个 id。

        用于 socket 连接 sid 与稳定 clientId 绑定后，把 session 从 sid 迁移到 clientId。

        规则：
        - 若 from_id 不存在：不做任何事
        - 若 to_id 已存在：合并 src->dst 的会话态后删除 from_id（避免因时序竞争导致开关/音色丢失）
        """
        if not from_id or not to_id or from_id == to_id:
            return

        src = self._sessions.get(from_id)
        if not src:
            return

        if to_id in self._sessions:
            dst = self._sessions.get(to_id)
            if dst is not None:
                self._merge_session_state(dst=dst, src=src)
                dst.session_id = to_id
            self._sessions.pop(from_id, None)
            return

        self._sessions[to_id] = src
        self._sessions.pop(from_id, None)
        src.session_id = to_id
        src.last_activity = datetime.utcnow()

    def get_or_create(self, sid: str) -> Session:
        if sid not in self._sessions:
            self._sessions[sid] = Session(session_id=sid)
        session = self._sessions[sid]
        session.last_activity = datetime.utcnow()
        return session

    def clear(self, sid: str) -> None:
        self._sessions.pop(sid, None)

    def add_message(self, sid: str, *, msg_type: str, content: str, metadata: Optional[Dict[str, Any]] = None) -> Dict[str, Any]:
        session = self.get_or_create(sid)
        record = {
            "id": f"msg_{int(time.time() * 1000)}_{sid}",
            "timestamp": datetime.utcnow().isoformat() + "Z",
            "type": msg_type,
            "content": content,
            "metadata": metadata or {},
        }
        session.history.append(record)
        if len(session.history) > 100:
            session.history = session.history[-100:]
        return record

    def get_history(self, sid: str, limit: int = 50) -> List[Dict[str, Any]]:
        session = self._sessions.get(sid)
        if not session:
            return []
        return session.history[-limit:]

    def set_pending_tool_call(self, sid: str, tool_call: Dict[str, Any], risk_assessment: Optional[Dict[str, Any]] = None) -> Dict[str, Any]:
        session = self.get_or_create(sid)
        pending = {
            **tool_call,
            "riskAssessment": risk_assessment,
            "createdAt": datetime.utcnow().isoformat() + "Z",
            "expiresAt": (datetime.utcnow() + self.confirmation_timeout).isoformat() + "Z",
        }
        session.pending_tool_call = pending
        session.canceled = False
        return pending

    def clear_pending_tool_call(self, sid: str) -> Optional[Dict[str, Any]]:
        session = self._sessions.get(sid)
        if not session:
            return None
        pending = session.pending_tool_call
        session.pending_tool_call = None
        session.canceled = False
        return pending

    def mark_canceled(self, sid: str, reason: str = "user_canceled") -> None:
        session = self.get_or_create(sid)
        session.canceled = True
        session.context["cancelReason"] = reason
        session.context["canceledAt"] = datetime.utcnow().isoformat() + "Z"

    def set_pending_confirmation(self, sid: str, confirmation: Dict[str, Any]) -> Dict[str, Any]:
        session = self.get_or_create(sid)
        session.pending_confirmation = {
            **confirmation,
            "createdAt": datetime.utcnow().isoformat() + "Z",
            "expiresAt": (datetime.utcnow() + self.confirmation_timeout).isoformat() + "Z",
        }
        return session.pending_confirmation

    def get_pending_confirmation(self, sid: str) -> Optional[Dict[str, Any]]:
        session = self._sessions.get(sid)
        if not session or not session.pending_confirmation:
            return None
        expires_at = session.pending_confirmation.get("expiresAt")
        if expires_at:
            try:
                exp = datetime.fromisoformat(expires_at.replace("Z", "+00:00"))
                if datetime.utcnow().replace(tzinfo=exp.tzinfo) > exp:
                    session.pending_confirmation = None
                    return None
            except Exception:
                # 若解析失败，为避免误伤用户，这里不阻断请求。
                pass
        return session.pending_confirmation

    def clear_pending_confirmation(self, sid: str) -> Optional[Dict[str, Any]]:
        session = self._sessions.get(sid)
        if not session:
            return None
        pending = session.pending_confirmation
        session.pending_confirmation = None
        return pending

    def get_status(self, sid: str) -> Dict[str, Any]:
        session = self._sessions.get(sid)
        if not session:
            return {"exists": False, "active": False}

        now = datetime.utcnow()
        is_expired = (now - session.last_activity) > self.session_timeout
        return {
            "exists": True,
            "active": not is_expired,
            "sessionId": session.session_id,
            "hasPendingToolCall": bool(session.pending_tool_call),
            "hasPendingConfirmation": bool(session.pending_confirmation),
            "isCanceled": bool(session.canceled),
            "messageCount": len(session.history),
            "lastActivity": session.last_activity.isoformat() + "Z",
            "createdAt": session.created_at.isoformat() + "Z",
        }

    def get_stats(self) -> Dict[str, Any]:
        return {
            "totalSessions": len(self._sessions),
            "activeSessions": len(self._sessions),
        }
